src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs

```diff
@@ -0,0 +1,367 @@
+                offset += bytesRead;
+            }
+
+            return await DecodeBufferedMessageAsync(new ArraySegment<byte>(buffer, 0, offset), inputStream);
+        }
+
+        private Message ReadStreamedMessage(Stream inputStream)
+        {
+            MaxMessageSizeStream maxMessageSizeStream = new MaxMessageSizeStream(inputStream, _factory.MaxReceivedMessageSize);
+
+            try
+            {
+                return _encoder.ReadMessage(maxMessageSizeStream, _factory.MaxBufferSize, _contentType);
+            }
+            catch (XmlException xmlException)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
+            }
+        }
+
+        private void ThrowMaxReceivedMessageSizeExceeded()
+        {
+            if (TD.MaxReceivedMessageSizeExceededIsEnabled())
+            {
+                TD.MaxReceivedMessageSizeExceeded(SR.Format(SR.MaxReceivedMessageSizeExceeded, _factory.MaxReceivedMessageSize));
+            }
+
+            string message = SR.Format(SR.MaxReceivedMessageSizeExceeded, _factory.MaxReceivedMessageSize);
+            Exception inner = new QuotaExceededException(message);
+            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new CommunicationException(message, inner));
+        }
+
+        private async Task<Message> DecodeBufferedMessageAsync(ArraySegment<byte> buffer, Stream inputStream)
+        {
+            try
+            {
+                // if we're chunked, make sure we've consumed the whole body
+                if (_contentLength == -1 && buffer.Count == _factory.MaxReceivedMessageSize)
+                {
+                    byte[] extraBuffer = new byte[1];
+                    int extraReceived = await inputStream.ReadAsync(extraBuffer, 0, 1);
+                    if (extraReceived > 0)
+                    {
+                        ThrowMaxReceivedMessageSizeExceeded();
+                    }
+                }
+
+                try
+                {
+                    return _encoder.ReadMessage(buffer, _factory.BufferManager, _contentType);
+                }
+                catch (XmlException xmlException)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                        new ProtocolException(SR.MessageXmlProtocolError, xmlException));
+                }
+            }
+            finally
+            {
+                inputStream.Dispose();
+            }
+        }
+
+        private async Task<Stream> GetStreamAsync()
+        {
+            var content = _httpResponseMessage.Content;
+            Stream contentStream = null;
+            _contentLength = -1;
+            if (content != null)
+            {
+                contentStream = await content.ReadAsStreamAsync();
+                _contentLength = content.Headers.ContentLength ?? -1;
```

Comment:
We may have to be careful about the use of the `??` operator as it's relatively new, and I recall that we may have to watch our uses of these newer keywords/operators. Please be sure this compiles in all the environments we need to work on prior to merging the PR.